drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1143 @@
+ *
+ * WARNING: Don't mix USB defined in STM32Cube HAL and CONFIG_USB_* from Zephyr
+ * Kconfig system.
+ */
+#define USB_NUM_BIDIR_ENDPOINTS	DT_INST_PROP(0, num_bidir_endpoints)
+
+#if defined(USB) || defined(USB_DRD_FS)
+#define EP0_MPS 64U
+#define EP_MPS 64U
+#define USB_BTABLE_SIZE  (8 * USB_NUM_BIDIR_ENDPOINTS)
+#define USB_RAM_SIZE	DT_INST_PROP(0, ram_size)
+#else /* USB_OTG_FS */
+#define EP0_MPS USB_OTG_MAX_EP0_SIZE
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
+#define EP_MPS USB_OTG_HS_MAX_PACKET_SIZE
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otgfs) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usb)
+#define EP_MPS USB_OTG_FS_MAX_PACKET_SIZE
+#endif
+#define USB_RAM_SIZE	DT_INST_PROP(0, ram_size)
+#define USB_BTABLE_SIZE 0
+#endif /* USB */
+
+#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
+			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
+#define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
+			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
+static struct udc_stm32_data udc0_priv;
+
+static struct udc_data udc0_data = {
+	.mutex = Z_MUTEX_INITIALIZER(udc0_data.mutex),
+	.priv = &udc0_priv,
+};
+
+static const struct udc_stm32_config udc0_cfg  = {
+	.num_endpoints = USB_NUM_BIDIR_ENDPOINTS,
+	.dram_size = USB_RAM_SIZE,
+	.pma_offset = USB_BTABLE_SIZE,
+	.ep0_mps = EP0_MPS,
+	.ep_mps = EP_MPS,
+};
+
+#if defined(USB_OTG_FS) || defined(USB_OTG_HS)
+static uint32_t usb_dc_stm32_get_maximum_speed(void)
+{
+/*
+ * STM32L4 series USB LL API doesn't provide HIGH and HIGH_IN_FULL speed
+ * defines.
+ */
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
+#define USB_OTG_SPEED_HIGH                     0U
+#define USB_OTG_SPEED_HIGH_IN_FULL             1U
+#endif /* CONFIG_SOC_SERIES_STM32L4X */
+/*
+ * If max-speed is not passed via DT, set it to USB controller's
+ * maximum hardware capability.
+ */
+#if USB_OTG_HS_EMB_PHY || USB_OTG_HS_ULPI_PHY
+	uint32_t speed = USB_OTG_SPEED_HIGH;
+#else
+	uint32_t speed = USB_OTG_SPEED_FULL;
+#endif
+
+#ifdef USB_MAXIMUM_SPEED
+
+	if (!strncmp(USB_MAXIMUM_SPEED, "high-speed", 10)) {
+		speed = USB_OTG_SPEED_HIGH;
+	} else if (!strncmp(USB_MAXIMUM_SPEED, "full-speed", 10)) {
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(USB_OTG_HS_EMB_PHY)
+		speed = USB_OTG_SPEED_HIGH_IN_FULL;
+#else
+		speed = USB_OTG_SPEED_FULL;
+#endif
+	} else {
+		LOG_DBG("Unsupported maximum speed defined in device tree. "
+			"USB controller will default to its maximum HW "
+			"capability");
+	}
+#endif
+
+	return speed;
+}
+#endif /* USB_OTG_FS || USB_OTG_HS */
+
+static void priv_pcd_prepare(const struct device *dev)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+
+	memset(&priv->pcd, 0, sizeof(priv->pcd));
+
+	/* Default values */
+	priv->pcd.Init.dev_endpoints = cfg->num_endpoints;
+	priv->pcd.Init.ep0_mps = cfg->ep0_mps;
+	priv->pcd.Init.speed = PCD_SPEED_FULL;
+	priv->pcd.Init.phy_itface = PCD_PHY_EMBEDDED;
```

Comment:
This value appears to be set twice in the same function

Reply:
@loicpoulain please check L:902

Reply:
fixed.